p2p/src/cybershake.rs

```diff
@@ -218,73 +240,255 @@ where
     Ok((received_remote_identity, outgoing, incoming))
 }
 
-// TODO: implement AsyncWrite for this, buffering the data and encrypting on flush or on each N-byte chunk.
+macro_rules! ready {
+    ($($tokens:tt)*) => {
+        match $($tokens)* {
+            Poll::Pending => return Poll::Pending,
+            Poll::Ready(Ok(n)) => { n }
+            Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
+        }
+    };
+}
+
 impl<W: AsyncWrite + Unpin> Outgoing<W> {
+    /// Send a message with any length.
     pub async fn send_message(&mut self, msg: &[u8]) -> Result<(), Error> {
+        let mut written = 0;
```

Comment:
Now that we decouple framing of messages from encryption, we probably can use something like Tokio Codecs to implement message framing outside of the cybershake module. See Henry's excellent write up on how to use codecs: https://www.zfnd.org/blog/decoding-bitcoin-messages-with-tokio-codecs/

Reply:
In this PR, though, we may keep this API and refactor it later.